fluture.js

```diff
@@ -376,544 +404,950 @@
-  Future.fork = createBinaryDispatcher('fork');
-  Future.race = createUnaryDispatcher('race');
-  Future.or = createUnaryDispatcher('or');
-  Future.fold = createBinaryDispatcher('fold');
-  Future.hook = createInvertedBinaryDispatcher('hook');
-  Future.finally = createUnaryDispatcher('finally');
-  Future.value = createUnaryDispatcher('value');
-  Future.promise = createNullaryDispatcher('promise');
-  Future.cache = createNullaryDispatcher('cache');
+  FutureParallel.prototype.toString = function FutureParallel$toString(){
+    return `Future.parallel(${show(this._max)}, [${this._futures.map(show).join(', ')}])`;
+  }
 
-  /////////////////////
-  // Other functions //
-  /////////////////////
+  //----------
 
-  Future.isFuture = isFuture;
-  Future.isForkable = isForkable;
+  function FutureDo(generator){
+    this._generator = generator;
+  }
 
-  Future.reject = function Future$reject(x){
-    return new FutureClass(false, function Future$reject$fork(rej){
-      rej(x);
-      return noop;
-    });
-  };
+  FutureDo.prototype = Object.create(Future.prototype);
 
-  Future.after = function Future$after(n, x){
-    if(arguments.length === 1) return unaryPartial(Future.after, n);
-    check$after(n);
-    return new FutureClass(false, function Future$after$fork(rej, res){
-      const t = setTimeout(res, n, x);
-      return function Future$after$cancel(){ clearTimeout(t) };
-    });
-  };
+  FutureDo.prototype._f = function FutureDo$fork(rej, res){
+    const iterator = this._generator();
+    check$do$g(iterator);
+    const recurser = new ChainRec(function Future$do$next(next, _, x){
+      const iteration = iterator.next(x);
+      check$do$next(iteration);
+      return iteration.done ? new FutureOf(iteration) : iteration.value.map(next);
+    }, undefined);
+    return recurser._f(rej, res);
+  }
 
-  Future.cast = function Future$cast(m){
-    check$cast(m);
-    return new FutureClass(true, function Future$cast$fork(rej, res){
-      m.fork(rej, res);
-    });
-  };
+  FutureDo.prototype.toString = function FutureDo$toString(){
+    return `Future.do(${showf(this._generator)})`;
+  }
 
-  Future.encase = function Future$encase(f, x){
-    check$encase(f);
-    if(arguments.length === 1) return unaryPartial(Future.encase, f);
-    return new FutureClass(false, function Future$encase$fork(rej, res){
+  //----------
+
+  function FutureCast(forkable){
+    SafeFuture.call(this, (l, r) => forkable.fork(l, r));
+    this._forkable = forkable;
+  }
+
+  FutureCast.prototype = Object.create(SafeFuture.prototype);
+
+  FutureCast.prototype.toString = function FutureCast$toString(){
+    return `Future.cast(${show(this._forkable)})`;
+  }
+
+  //----------
+
+  function FutureTry(fn){
+    this._fn = fn;
+  }
+
+  FutureTry.prototype = Object.create(Future.prototype);
+
+  FutureTry.prototype._f = function FutureTry$0$fork(rej, res){
+    let r;
+    try{ r = this._fn() }catch(e){ rej(e); return noop }
+    res(r);
+    return noop;
+  }
```

Comment:
We can't have the call to `res` reside inside the `try` block, as it will cause any subsequent actions triggered by calling `res` to also have their thrown Errors caught, causing the Future to enter the rejection branch after already being resolved... [scary stuff](https://github.com/Avaq/Fluture/commit/e9006a8606dc9f635c564a7d570229eb4c4255df).